torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,536 @@
+//
+// A very important concept in number theory is "congruence", as described
+// below:
+//
+// Definition 3.1: For c != 0, if (a-b)\c is an integer, then we say a is
+// congruent to b modulo c, written as a = b (mod c).
+//
+// Note that the "(mod k)" in the statement "a = b (mod k)" should be understood
+// as a qualifier for the statement "a = b" that modifies the meaning of the
+// original statement [P] to "[P] under the modular arithmetic". It should NOT
+// be understood as an operator operating on "b", i.e. "a = (b (mod k))".
+//
+// Theorem 3.2: The congruence is an equivalence relationship, that is:
+// - a = a (mod c)
+// - if a = b (mod c), then b = a (mod c)
+// - if a = b (mod c) and b = d (mod c), then a = d (mod c)
+//
+// Theorem 3.3: The congruence also has the following properties:
+// - a = b (mod c) iff a = b (mod (-c))
+// - if a = a' (mod c), b = b' (mod c), then
+//   - a + b = a' + b' (mod c)
+//   - ab = a'b' (mod c)
+// - if bd = bd' (mod c) and gcd(b, c) = 1, then d = d' (mod c)
+//
+// 4. Some More Theorems
+// =====================
+//
+// In this section, I will prove a few theorems that I didn't find in textbooks,
+// but still feel useful for us. The / and % are under the definition of number
+// theory, not under the definition of C++. The properties of / and % with the
+// definition of C++ will be revisited in later section "Properties of Div and
+// Mod Under Trunc Div".
+//
+// Bad news comes first:
+//
+// Theorem 4.1: Integer div is not associative:
+// 1) a*(b/c) != (a*b)/c
+// 2) a/(b*c) != (a/b)/c
+// 3) a/(b/c) != (a/b)*c != (a*c)/b
+// Proof: a*(b/c) != (a*b)/c because of the counter example 2*(3/2) != (2*3)/2
+// a/(b*c) != (a/b)/c because of the counter example 4/((-1)*5) != (4/(-1))/5
+// Note that 4/(-5) = 0, (-4)/5 = -1
+// a/(b/c) != (a/b)*c != (a*c)/b because of the counter example that
+// 5/(3/2), (5/3)*2, and (5*2)/3 are three different numbers
+//
+// Theorem 4.2: Integer div is NOT right distributive
+// Proof: Counter example: (1+1)/2 != 1/2 + 1/2
+//
+// Then comes good properties:
+//
+// Theorem 4.3: a%b = a'%b is equivalent to a = a' (mod b)
+// Proof: Direction ==>:
+// a%b = a'%b is equivalent to a-(a/b)b = a'-(a'/b)b
+// which is equivalent to (a-a')\b = (a/b-a'/b) = integer
+// So a%b = a'%b ==> a = a' (mod b)
+// Direction <==:
+// if a = a' (mod b), then a = a' + kb.
+// According to the Euclid's division lemma, a' = q'b + r',
+// then a = (q'+k)b + r', where q'+k and r'. It is easy to verify that, if we
+// define q = q'+k and r = r', then a = qb + r also satisfies the condition in
+// Euclid's division lemma. Thanks to the uniqueness of q and r, we have
+// a%b = r = r' = a'%b
+//
+// Thanks to the great property of Theorem 4.3 some theorems below can be easily
+// proved by converting % into congruence. But unfortunately, as we can see in
+// section "Implementations of Div and Mod", for the trunc div in C++, the
```

Comment:
This I'm curious to see because, per the above comment, I would expect C++ to preserve this property

Reply:
Unfortunately not. I just added a counter example in section 6:
```C++
// For trunc div, Theorem 4.3 no longer holds, because -3 = 2 (mod 5), however,
// -3 % 5 = -3, but 2 % 5 = 2.
```